sygnal/gcmpushkin.py

```diff
@@ -47,197 +56,369 @@
-    'InvalidRegistration',
-    'NotRegistered',
-    'InvalidPackageName',
-    'MismatchSenderId',
+    "MissingRegistration",
+    "InvalidRegistration",
+    "NotRegistered",
+    "InvalidPackageName",
+    "MismatchSenderId",
 ]
 
 # Failure codes that mean the message in question will never
 # succeed, so don't retry, but the registration ID is fine
 # so we should not reject it upstream.
-BAD_MESSAGE_FAILURE_CODES = [
-    'MessageTooBig',
-    'InvalidDataKey',
-    'InvalidTtl',
-]
+BAD_MESSAGE_FAILURE_CODES = ["MessageTooBig", "InvalidDataKey", "InvalidTtl"]
+
+DEFAULT_MAX_CONNECTIONS = 20
+
 
 class GcmPushkin(Pushkin):
+    """
+    Pushkin that relays notifications to Google/Firebase Cloud Messaging.
+    """
+
+    UNDERSTOOD_CONFIG_FIELDS = {"type", "api_key"}
 
-    def __init__(self, name):
-        super(GcmPushkin, self).__init__(name)
-        self.session = Session()
-        a = adapters.HTTPAdapter(pool_maxsize=20, pool_connections=20, pool_block=True)
-        self.session.mount("https://", a)
+    def __init__(self, name, sygnal, config):
+        super(GcmPushkin, self).__init__(name, sygnal, config)
 
-    def setup(self, ctx):
-        self.db = ctx.database
+        nonunderstood = set(self.cfg.keys()).difference(self.UNDERSTOOD_CONFIG_FIELDS)
+        if len(nonunderstood) > 0:
+            logger.warning(
+                "The following configuration fields are not understood: %s",
+                nonunderstood,
+            )
 
-        self.api_key = self.getConfig('apiKey')
+        self.http_agent = None
+        self.http_pool = None
+        self.db = None
+        self.canonical_reg_id_store = None
+
+        self.api_key = self.get_config("api_key")
         if not self.api_key:
             raise PushkinSetupException("No API key set in config")
-        self.canonical_reg_id_store = CanonicalRegIdStore(self.db)
-
-    def dispatchNotification(self, n):
-        pushkeys = [device.pushkey for device in n.devices if device.app_id == self.name]
-        # Resolve canonical IDs for all pushkeys
-        pushkeys = [canonical_reg_id or reg_id for (reg_id, canonical_reg_id) in
-                    self.canonical_reg_id_store.get_canonical_ids(pushkeys).items()]
 
-        data = GcmPushkin.build_data(n)
-        headers = {
-            "User-Agent": "sygnal",
-            "Content-Type": "application/json",
-            "Authorization": "key=%s" % (self.api_key,)
-        }
+    async def start(self, sygnal):
+        self.http_pool = HTTPConnectionPool(reactor=sygnal.reactor)
+        self.http_pool.maxPersistentPerHost = self.get_config(
+            "max_connections", DEFAULT_MAX_CONNECTIONS
+        )
+
+        self.http_agent = Agent(reactor=sygnal.reactor, pool=self.http_pool)
+
+        self.db = sygnal.database
+
+        logger.debug("About to set up CanonicalRegId Store")
+        self.canonical_reg_id_store = CanonicalRegIdStore()
+        await self.canonical_reg_id_store.setup(self.db)
```

Comment:
I'd rather we passed the store in via the constructor, that way a) `GcmPushkin` always has a valid store, b) we can then fold the start methods for the pushkins into the init as they no longer have to do async stuff which will make `run` a *lot* simpler.

Reply:
This might be a struggle since the pushkin is constructed based on import name with a standard set of arguments that would be used to construct any pushkin. As such, passing it in the constructor would need adding special casing to ``.

We should also consider third-party pushkins which may benefit from being able to run async code at startup.

Thoughts? @erikjohnston

Reply:
> This might be a struggle since the pushkin is constructed based on import name with a standard set of arguments that would be used to construct any pushkin. As such, passing it in the constructor would need adding special casing for GCM.

We only have two types of pushkins so special casing is far from the end of world, there's something to be said for simplicity here. The other option is to add the store to the `Sygnal` object and then the `GcmPushkin` can grab the store from that. 

> We should also consider third-party pushkins which may benefit from being able to run async code at startup.

TBH dealing with third party pushkins is something to worry about when it happens.

--- 

Broadly I'm not a fan of having two steps to instansiate an object, it tends to lead to footguns where either the `setup` functions isn't called or is called more than once. Ideally once you have an instantiated object then it should be ready to be used, which basically means having to pass in everything the class needs into the constructor. If you really need some async setup then you probably want a static async method that does the work and then instantiates and returns the object.

Reply:
I think my preferred solution then would be to have a static async method to construct everything. This gives the most flexibility without letting GCM-specific stuff (CanonicalRegIdStore) leech out of `gcmpushkin.py`.

I have already had 2 enquiries about custom pushkins, so it is probably worth offering the flexibility now. The code has been in place since last Sygnal, it may as well be left in.
Calling `async` code from a synchronous context can be a bit of a dog, I think it's worth leaving it open as an option. In the event that it isn't needed, nothing save from having to write 6 characters (`async `) is lost.